sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -0,0 +1,347 @@
+            },
+            SyntaxKind.MethodDeclaration);
+        }
+
+        public IEnumerable<Diagnostic> FindViolations(SyntaxNodeAnalysisContext context)
+        {
+            var methodSymbol = context.SemanticModel.GetDeclaredSymbol(context.Node) as IMethodSymbol;
+
+            if (methodSymbol == null ||
+                methodSymbol.Parameters.Length == 0 ||
+                methodSymbol.IsOverride)
+            {
+                return Enumerable.Empty<Diagnostic>();
+            }
+
+            var parametersToCheck = new Dictionary<string, ParameterData>();
+            foreach (var parameter in methodSymbol.Parameters.Where(IsTrackedParameter))
+            {
+                parametersToCheck[parameter.Name] = new ParameterData(parameter);
+            }
+
+            var parameterUsesInMethod = context
+                .Node
+                .DescendantNodes()
+                .OfType<IdentifierNameSyntax>()
+                .Where(id => parametersToCheck.Values.Any(p => p.MatchesIdentifier(id, context.SemanticModel)));
+
+            foreach (var identifierReference in parameterUsesInMethod)
+            {
+                var key = identifierReference.Identifier.ValueText ?? "";
+
+                ParameterData paramData;
+                if (!parametersToCheck.TryGetValue(key, out paramData) ||
+                    !paramData.ShouldReportOn)
+                {
+                    continue;
+                }
+
+                if (identifierReference.Parent is EqualsValueClauseSyntax ||
+                    identifierReference.Parent is AssignmentExpressionSyntax)
+                {
+                    paramData.ShouldReportOn = false;
+                    continue;
+                }
+
+                var symbolUsedAs = FindParameterUseAsType(identifierReference, context.SemanticModel);
+                if (symbolUsedAs != null)
+                {
+                    paramData.AddUsage(symbolUsedAs);
+                }
+            }
+
+            return parametersToCheck
+                .Values
+                .Select(p => p.GetRuleViolation())
+                .WhereNotNull();
+        }
+
+        private static bool IsTrackedParameter(IParameterSymbol parameterSymbol)
+        {
+            var type = parameterSymbol.Type;
+
+            return !type.DerivesFrom(KnownType.System_Array) &&
+                   !type.IsValueType &&
+                   !type.Is(KnownType.System_String);
+        }
+
+        private SyntaxNode GetNextUnparethesizedParent(SyntaxNode node)
+        {
+            var expression = node as ExpressionSyntax;
+            if (expression == null)
+            {
+                return node;
+            }
+
+            var topmostParent = SyntaxHelper.GetSelfOrTopParenthesizedExpression(expression);
+
+            return (topmostParent is ParenthesizedExpressionSyntax)
+                ? topmostParent.Parent
+                : topmostParent;
```

Comment:
Looks weird. Your method suggest you are looking for the parent but here you don't.

Examples:
- with `foo.Bar`, topmostParent will return (when called on foo) foo
- with `(((foo))).Bar`, topmostParent will return (when called on foo) the outer parenthesis expresion

So in both cases you need to call `.Parent` to access to the `MemberAccessExpression`.